### ISSUE TITLE
Jolokia input plugin should extract tags

### ISSUE THREAD

**Issue by dylanmei** [influxdata/telegraf#2014]
## Feature Request

Jolokia input plugin should have a mechanism for extracting tags to account for an evolving application. If you have a workaround, or think I'm simply doing it wrong, please share.

### Proposal:

The Jolokia input plugin is well suited for a fixed set of `mbean` names. However, in a system such as Kafka or Cassandra where metrics may come and go (based on user created/deleted topics, tables, and so on) the input plugin is not adaptive enough. I'd like to capture values such as topics and tables as tags, which keeps measurements simple and querying expressive.

### Current behavior:

If I follow what appears to be the default pattern of setting up Jolokia, I get this kind of measurement:

```
jolokia,host=myhost \
  kafka_server_kafka.server:name\=MessagesInPerSec\,topic\=foo-logs\,type\=BrokerTopicMetrics_Count=1,23 \
  kafka_server_kafka.server:name\=MessagesInPerSec\,topic\=foo-logs\,type\=BrokerTopicMetrics_EventType="messages", \
  kafka_server_kafka.server:name\=MessagesInPerSec\,topic\=foo-logs\,type\=BrokerTopicMetrics_FifteenMinuteRate=123, \
  kafka_server_kafka.server:name\=MessagesInPerSec\,topic\=foo-logs\,type\=BrokerTopicMetrics_FiveMinuteRate=123, \
  kafka_server_kafka.server:name\=MessagesInPerSec\,topic\=foo-logs\,type\=BrokerTopicMetrics_MeanRate=123, \
  kafka_server_kafka.server:name\=MessagesInPerSec\,topic\=foo-logs\,type\=BrokerTopicMetrics_OneMinuteRate=123, \
  kafka_server_kafka.server:name\=MessagesInPerSec\,topic\=foo-logs\,type\=BrokerTopicMetrics_RateUnit="SECONDS" \
  1478618057000000000
```

...given this config:

```
[[inputs.jolokia]]
  # omitted

  [[inputs.jolokia.servers]]
    # omitted

  [[inputs.jolokia.metrics]]
    name   = "messages_in_per_sec"
    mbean = "kafka.server:type=BrokerTopicMetrics,topic=*,name=MessagesInPerSec"
```

The more topics I add to the cluster, the ***larger that measurement grows***. 

To get Kafka measurements per-topic like this:

```
kafka_server,host=myhost,topic=foo-logs \
  messages_in_per_sec_Count=123, \
  messages_in_per_sec_FifteenMinuteRate=123, \
  messages_in_per_sec_FiveMinuteRate=123, \
  messages_in_per_sec_MeanRate=123, \
  messages_in_per_sec_OneMinuteRate=123 \
  1478618057000000000
```
...requires a new Jolokia input per Kafka topic.

```
[[inputs.jolokia]]
  name_override = "kafka_server"

  [[inputs.jolokia.servers]]
    # omitted

  [inputs.jolokia.tags]
    topic = "foo-logs"

  [[inputs.jolokia.metrics]]
    name      = "messages_in_per_sec"
    mbean    = "kafka.server:type=BrokerTopicMetrics,topic=foo-logs,name=MessagesInPerSec"
    attribute = "Count,OneMinuteRate,FiveMinuteRate,FifteenMinuteRate,MeanRate"

```

The former config is better because it picks up new topics automatically with fewer requests, but the latter config creates measurements that scale and, to my eye, are more succinct.

### Desired behavior:

To get the best of both worlds, the Jolokia plugin could extract tags from the results that are returned. Using, for example, Logstash Grok capture, given this mbean specification: `kafka.server:name=MessagesInPerSec,topic=(?<topic>[^,]+),type=BrokerTopicMetrics` we could extract `topic` tags for this Jolokia response:

```
{
  "request": {
    "mbean": "kafka.server:name=MessagesInPerSec,topic=*,type=BrokerTopicMetrics",
    "type": "read"
  },
  "value": {
    "kafka.server:name=MessagesInPerSec,topic=foo-logs,type=BrokerTopicMetrics": {
      ...
    },
    "kafka.server:name=MessagesInPerSec,topic=other-logs,type=BrokerTopicMetrics": {
      ...
    }
  }
}
```

### Use case: [Why is this important (helps with prioritizing requests)]

What we want is:
- fewer Jolokia requests
- avoid configuration changes just to account for application state
- adapt to new mbean names with tags instead of fields

**Comment by sparrc** [comment]
this is a good idea, I agree that we can do a better job at parsing mbean-formatted metrics, logstash would be a good model to follow.

**Comment by dylanmei** [comment]
@sparrc I've looked at this quite a bit, and it can be done but with some major changes to the Jolokia plugin.

After reading issue #1726, I felt that another way to solve this problem is by transforming the unwieldy measurements that Jolokia returns, essentially extracting tags from fields.

What do you think about:

1. using a processor plugin for this use case?
2. is this already possible and I just haven't found the plugin?

**Comment by sparrc** [comment]
what would be the changes to the jolokia plugin? would they make the plugin better for everyone?

I'm sure a processor plugin could do it, but it seems like we should do our best to make the jolokia plugin as usable as possible.

TBH I'm not familiar with jolokia but the plugin as it stands seems to draw a lot of ire....

**Comment by sparrc** [comment]
cc @phemmer

**Comment by mainvoid** [comment]
Would #2242 help?

It does the same thing and is much more friendly to work with.

**Comment by sparrc** [comment]
for the lazy here's the cassandra plugin readme: https://github.com/influxdata/telegraf/tree/master/plugins/inputs/cassandra

**Comment by dylanmei** [comment]
Thank you for your replies!

@sparrc For me, my "ire" comes from how the Jolokia input plugin handles the result of a wildcard read request: [section 6.2.1.3 Read response](https://jolokia.org/reference/html/protocol.html#read). The behavior is enshrined in this unit test, where fields have the mbean name encoded in them: [jolokia_test.go](https://github.com/influxdata/telegraf/blob/b7d29ca0e94cb9a89af16f313aefd5d19d5fc002/plugins/inputs/jolokia/jolokia_test.go#L139)

The Cassandra plugin accounts for this behavior, so instead of a Jolokia input plugin measurement like:

```
jolokia,host=myhost cassandraTable_org.apache.cassandra.metrics:keyspace=foo,name=ReadLatency,scope=bar,type=Table_Count=123
```

which I challenge anyone to write readable InfluxQL with, we get the Cassandra input plugin measurement:

```
cassandraTable,host=myhost,keyspace=foo,scope=bar ReadLatency_Count=123
```

Which is lovely except now the guy that wants to do the same thing with Kafka metrics (me) is copy-pasting the Cassandra plugin because the Jolokia plugin's wildcard behavior is naive.

Without breaking the current wildcard behavior, I was considering adding a config like `extract_tags_from_wildcards` so that, given this config:

```
[inputs.jolokia]]
  name_override = "cassandraTable"

  [[inputs.jolokia.servers]]

  [[inputs.jolokia.metrics]]
    name   = "ReadLatency"
    mbean = "org.apache.cassandra.metrics:keyspace=*,name=ReadLatency,scope=*,type=Table"
    extract_tags_from_wildcards = true
```

we could produce:

```
cassandraTable,host=myhost,keyspace=foo,scope=bar ReadLatency_Count=123
```

BTW, Jolokia is smart enough to know that `mbean = "org.apa*Table"` is invalid; wildcards are thankfully associated with name/value pairs.

Finally, I think you can create canned Cassandra and Kafka inputs around this functionality if it existed.

Back to the processor, a Grok/Mutate processor would be an interesting path to solving the problem, but would not help make the Jolokia plugin better.

**Comment by sparrc** [comment]
It seems like the way the cassandra metrics are represented is much cleaner. I wish that I had known this earlier and I probably wouldn't have accepted the jolokia plugin the way it is in the first place.

I wouldn't be opposed to breaking backwards compatibility in this case and going with the "cassandra way"

does anyone have strong attachments to the current state of jolokia metrics?

**Comment by dylanmei** [comment]
Hey, it got us this far. I believe `extract_tags_from_wildcards` is the way to go.

The Cassandra plugin could leverage the feature to call the Jolokia plugin like a library. We could make `Java` and `Kafka` plugins that are also thin wrappers around the Jolokia plugin. Only the brave and stubborn would then use it directly.

**Comment by sparrc** [comment]
We could do that, but are there specific Kafka metrics that can not be captured by the jolokia plugin?

I'd prefer if still had a generic `jolokia` solution that actually behaves nicely. We could always make a `jolokia2` and plan on deprecating the current `jolokia` plugin in favor of that?

**Comment by dylanmei** [comment]
I wasn't suggesting to get rid of the Jolokia plugin, only adding this feature to it. Then refactoring any Jolokia-based plugins like Cassandra to use the internals of it.

The Cassandra plugin basically does everything the Jolokia plugin does. Some of it is copy-paste from Jolokia, but with the benefit of context it can make prettier metrics.

Now @phemmer has added bulk-read to Jolokia, but Cassandra won't get it obviously.

**Comment by phemmer** [comment]
Since i've been pinged a few times in here, I'll provide my feedback.

> does anyone have strong attachments to the current state of jolokia metrics?

No. I do not use the current wildcard behavior, and doubt very many do (for the reasons already mentioned).

Other thoughts to consider:
## Wildcard in mbean domain
Solr exports metrics using the core name as part of the mbean domain. For example:

    solr/MC_1_CatalogEntry:id=org.apache.solr.handler.admin.ShowFileRequestHandler,type=/admin/file

It would be useful to be able to wildcard the mbean domain, and get it as a tag, so that we don't have to hard-code the solr core names. And even better would be to just get the part after the `/` as the tag value. The catch is that there's obviously no key=value here, so we'd need some alternate method for defining the tag name there. We could just use the tag `domain` when a wildcard is found in the domain. But if I do `solr/*:id=org.apache.solr.handler...", should I get `domain=solr/MC_1_CatalogEntry` or `domain=MC_1_CatalogEntry`?

## Multiple measurements per input
I think we need a way to generate multiple measurement names. Going back to solr again, there are a lot of handlers

```
solr/MC_1_CatalogEntry:id=org.apache.solr.handler.DumpRequestHandler,type=/debug/dump
solr/MC_1_CatalogEntry:id=org.apache.solr.handler.GraphHandler,type=/graph
solr/MC_1_CatalogEntry:id=org.apache.solr.handler.admin.SegmentsInfoRequestHandler,type=/admin/segments
solr/MC_1_CatalogEntry:id=org.apache.solr.handler.admin.ShowFileRequestHandler,type=/admin/file
...
```

But all the handlers are very similar, and so it makes sense to put them all into the same measurement, using tags to differentiate them.
However I would not want to put something like heap metrics, or solr's caching metrics, into the same measurement, as the fields are completely different.

Thus this might not work well with the earlier proposal of

```
  [[inputs.jolokia.metrics]]
    name   = "ReadLatency"
    mbean = "org.apache.cassandra.metrics:keyspace=*,name=ReadLatency,scope=*,type=Table"
```
generating
```
cassandraTable,host=myhost,keyspace=foo,scope=bar ReadLatency_Count=123
```

In my scenario, it would instead make sense to use the metric name in the measurement name, resulting in `cassandraTable_ReadLatency`

I could then do
```
[[inputs.jolokia.metrics]]
  name = "handlers"
  mbean = "solr/*:id=org.apache.solr.handler.*,type=*"
  attribute = "requests"
[[inputs.jolokia.metrics]]
  name = "memory"
  mbean = "java.lang:type=MemoryPool,name=*"
  attribute = "Usage,UsageThreshold"
```

And get:

```
solr_handlers,domain=MC_1_CatalogEntry,id=admin.ShowFileRequestHandler,type=/admin/file requests=123i
solr_handlers,domain=MC_1_CatalogEntry,id=GraphHandler,type=/graph requests=456i
solr_handlers,domain=MC_1_CatalogEntry,id=DumpRequestHandler,type=/debug/dump requests=789i
solr_memory,name="CMS Old Gen" Usage=111i,UsageThreshold=222i
solr_memory,name="Code Cache" Usage=333i,UsageThreshold=444i
```

**Comment by dylanmei** [comment]
Adding `domain` auto-magically as a tag is an interesting idea.

```
% curl -s "http://server:8778/jolokia/read/*:type=Runtime/Uptime,StartTime" | jq
{
  "request": {
    "mbean": "*:type=Runtime",
    "attribute": ["Uptime",  "StartTime"],
    ...
  },
  "value": {
    "java.lang:type=Runtime": {
      "Uptime": 94244283,
      "StartTime": 1484164511172
    }
  },
  ...
}
```

In the above case, using the "extract wildcard as tag idea", the name of the tag might be pre-defined as "domain" and the value is "java.lang".

> In my scenario, it would instead make sense to use the metric name in the measurement name, resulting in `cassandraTable_ReadLatency`

Achieving that would generate a lot of small measurements. But for the `solr_handlers` example, simply use `name_override` to rename `jolokia` to `solr`.

**Comment by phemmer** [comment]
> Achieving that would generate a lot of small measurements. But for the solr_handlers example, simply use name_override to rename jolokia to solr.

This doesn't work, as every `[inputs.jolokia.metrics]` would get the same measurement name.

**Comment by dylanmei** [comment]
Doh. Yes, that is right, because the plugin doesn't work how my intuition expects it to work.

**Comment by phemmer** [comment]
Oh, and

> Achieving that would generate a lot of small measurements.

I don't think it would. You can just use the same `name` parameter on multiple `[inputs.jolokia.metrics]`, and they'll all go into the same measurement.
The only conflict with your proposal is that the `name` would get used in the measurement name, not the field name. I would argue that this is the more natural behavior anyway, as doing it with a field prefix leads to having multiple points with different fields in the same measurement (which is counter to influxdb's design). However if you do want to add a prefix to all fields, what about a `field_prefix` parameter. You could then set it to `field_prefix = "ReadLatency_"`

**Comment by dylanmei** [comment]
@phemmer Would a config such as this account for the measurement name concern you have?

```
[[inputs.jolokia]]
  #...
  [[inputs.jolokia.servers]]
    #...

    [[inputs.jolokia.measurements]]
      name = "cassandraTable_ReadLatency"

      [[inputs.jolokia.metrics]]
        field_prefix = ""
        mbean = "org.apache.cassandra.metrics:keyspace=*,name=ReadLatency,scope=*,type=Table"
```

output: `cassandraTable_ReadLatency,host=myhost,keyspace=foo,scope=bar Count=123`

**Comment by danielnelson** [comment]
I believe this can be closed now that we have merged #2278.
